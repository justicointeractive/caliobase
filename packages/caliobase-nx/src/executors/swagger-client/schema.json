{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "title": "SwaggerClient executor",
  "description": "",
  "type": "object",
  "properties": {
    "input": {
      "type": "string",
      "description": "path to an openapi/swagger spec file"
    },
    "output": {
      "type": "string",
      "description": "path to the directory where the generated client will be saved"
    },
    "generateSpecTarget": {
      "type": "string",
      "description": "a target to be executed that should generate a swagger file"
    },
    "dotenv": {
      "type": "string",
      "description": "path to a dotenv file to load vars from"
    }
  },
  "required": []
}
