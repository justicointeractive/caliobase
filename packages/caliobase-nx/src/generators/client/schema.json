{
  "$schema": "http://json-schema.org/schema",
  "$id": "Client",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "apiProjectName": {
      "type": "string",
      "description": "",
      "x-prompt": "Which project provides the backend api?",
      "x-dropdown": "projects"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed",
      "alias": "d"
    }
  },
  "required": [
    "name",
    "apiProjectName"
  ]
}
