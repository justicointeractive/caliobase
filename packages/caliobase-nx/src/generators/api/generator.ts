import {
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  Tree,
} from '@nrwl/devkit';
import { applicationGenerator } from '@nrwl/nest/src/generators/application/application';
import * as path from 'path';
import { ApiGeneratorSchema } from './schema';

interface NormalizedSchema extends ApiGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: ApiGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).appsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

export default async function (tree: Tree, options: ApiGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  await applicationGenerator(tree, { name: options.name });
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    normalizedOptions.projectRoot,
    {
      cmsProjectName: null,
      ...options,
      ...normalizedOptions,
      tmpl: '',
    }
  );
  await formatFiles(tree);
}
