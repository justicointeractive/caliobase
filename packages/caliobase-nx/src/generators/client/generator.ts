import {
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  Tree,
} from '@nx/devkit';
import { libraryGenerator } from '@nx/js/src/generators/library/library';
import * as path from 'path';
import { modifyProjectConfiguration } from '../../lib/modifyProjectConfiguration';
import { ClientGeneratorSchema } from './schema';

interface NormalizedSchema extends ClientGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: ClientGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).libsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    tmpl: '',
  };
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    options.projectRoot,
    templateOptions
  );
}

export default async function (tree: Tree, options: ClientGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  await libraryGenerator(tree, {
    name: options.name,
    directory: options.directory,
    skipFormat: true,
  });
  addFiles(tree, normalizedOptions);
  modifyProjectConfiguration(tree, options.name, (config) => {
    config.targets = {
      ...config.targets,
      'generate-sources': {
        executor: '@caliobase/caliobase-nx:swagger-client',
        outputs: [`${normalizedOptions.projectRoot}/src`],
        options: {
          input: `./${normalizedOptions.projectRoot}/assets/openapi.json`,
          output: `./${normalizedOptions.projectRoot}/src/lib`,
          name: `${normalizedOptions.projectName}.ts`,
        },
        dependsOn: [{ projects: 'dependencies', target: 'swagger' }],
      },
    };
    config.implicitDependencies = [
      ...(config.implicitDependencies ?? []),
      options.apiProjectName,
    ];

    return config;
  });
  modifyProjectConfiguration(
    tree,
    options.apiProjectName,
    (apiProjectConfig) => {
      apiProjectConfig.targets = {
        ...apiProjectConfig.targets,
        swagger: {
          executor: 'nx:run-commands',
          outputs: [`${normalizedOptions.projectRoot}/assets`],
          dependsOn: [
            {
              target: 'build',
              projects: 'self',
            },
          ],
          options: {
            command: `node ./dist/${apiProjectConfig.root}/main.js --write-swagger-and-exit=./${normalizedOptions.projectRoot}/assets/openapi.json`,
          },
        },
      };

      return apiProjectConfig;
    }
  );

  // TODO: update nx.json's targetDefaults build dependsOn "^generate-sources", "generate-sources"

  await formatFiles(tree);
}
