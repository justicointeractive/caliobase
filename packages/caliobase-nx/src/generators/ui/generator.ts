import {
  formatFiles,
  generateFiles,
  GeneratorCallback,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  Tree,
} from '@nrwl/devkit';
import { Linter } from '@nrwl/linter';
import { applicationGenerator } from '@nrwl/react';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import * as path from 'path';
import { modifyProjectConfiguration } from '../../lib/modifyProjectConfiguration';
import { addDependencyVersionsToPackageJson } from '../../lib/versions';
import { UiGeneratorSchema } from './schema';
import assert = require('assert');

interface NormalizedSchema extends UiGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: UiGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).appsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    template: '',
  };
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    options.projectRoot,
    templateOptions
  );
}

export default async function (tree: Tree, options: UiGeneratorSchema) {
  const tasks: GeneratorCallback[] = [];
  const normalizedOptions = normalizeOptions(tree, options);
  await applicationGenerator(tree, {
    ...options,
    e2eTestRunner: 'cypress',
    linter: Linter.EsLint,
    unitTestRunner: 'jest',
    skipFormat: true,
    style: 'css',
  });
  addFiles(tree, normalizedOptions);
  modifyProjectConfiguration(tree, options.name, (config) => {
    assert(config.targets);
    config.targets['serve'].options[
      'proxyConfig'
    ] = `${normalizedOptions.projectRoot}/proxy.conf.json`;
    return config;
  });
  await formatFiles(tree);

  tasks.push(
    addDependencyVersionsToPackageJson(tree, ['@caliobase/caliobase-ui'])
  );

  return runTasksInSerial(...tasks);
}
