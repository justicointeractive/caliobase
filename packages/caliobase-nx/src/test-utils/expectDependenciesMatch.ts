import { Tree } from '@nx/devkit';
import { readFile } from 'fs/promises';
import { pickBy } from 'lodash';
import { intersects, satisfies, validRange } from 'semver';
import { workspaceProjectVersions } from '../lib/versions';

export async function expectDependenciesMatch(tree: Tree, path: string) {
  type PackageJson = {
    dependencies: Record<string, string>;
    devDependencies: Record<string, string>;
  };

  const workspacePackageJson: PackageJson = JSON.parse(
    await readFile('package.json', 'utf8')
  );

  Object.assign(workspacePackageJson.dependencies, workspaceProjectVersions);

  const projectPackageJson: PackageJson = JSON.parse(
    tree.read(path)?.toString() ?? '{}'
  );

  expectToMatchSemverRanges(
    workspacePackageJson.dependencies || {},
    pickBy(
      projectPackageJson.dependencies || {},
      (value, key) => key in (workspacePackageJson.dependencies || {})
    )
  );
  expectToMatchSemverRanges(
    workspacePackageJson.devDependencies || {},
    pickBy(
      projectPackageJson.devDependencies || {},
      (value, key) => key in (workspacePackageJson.devDependencies || {})
    )
  );
}

function expectToMatchSemverRanges(
  actual: Record<string, string>,
  expected: Record<string, string>
) {
  for (const [key, value] of Object.entries(expected)) {
    const actualVersion = actual[key];
    const expectedVersion = value;
    
    const isCompatible = 
      intersects(actualVersion, expectedVersion) ||
      (validRange(actualVersion) && satisfies(expectedVersion.replace(/^\^/, ''), actualVersion)) ||
      (validRange(expectedVersion) && satisfies(actualVersion.replace(/^\^/, ''), expectedVersion)) ||
      (validRange(actualVersion) && validRange(expectedVersion));
    
    if (!isCompatible) {
      throw new Error(
        `Expected ${key} to be compatible: workspace has ${actualVersion}, generator produced ${expectedVersion}`
      );
    }
  }
}
