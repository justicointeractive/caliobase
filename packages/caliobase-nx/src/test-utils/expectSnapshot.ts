import { Tree } from '@nrwl/devkit';

export function expectSnapshot(tree: Tree, path: string) {
  expect(tree.read(path)?.toString()).toMatchSnapshot(path);
}
