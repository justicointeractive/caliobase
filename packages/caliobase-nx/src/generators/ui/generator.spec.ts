import { readProjectConfiguration, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { expectSnapshot } from '../../test-utils/expectSnapshot';
import generator from './generator';
import { UiGeneratorSchema } from './schema';

describe('ui generator', () => {
  let appTree: Tree;
  const options: UiGeneratorSchema = {
    name: 'ui',
  };

  beforeEach(async () => {
    appTree = createTreeWithEmptyWorkspace();
  });

  it('should run successfully', async () => {
    await generator(appTree, options);
    const config = readProjectConfiguration(appTree, 'ui');
    expect(config).toBeDefined();
    expectSnapshot(appTree, 'package.json', [
      [
        /"@caliobase\/caliobase-ui": "[0-9.]+"/g,
        '"@caliobase/caliobase-ui": "*"',
      ],
    ]);
    expectSnapshot(appTree, 'apps/ui/project.json');
    expectSnapshot(appTree, 'apps/ui/src/main.tsx');
    expectSnapshot(appTree, 'apps/ui/src/styles.css');
  });
});
