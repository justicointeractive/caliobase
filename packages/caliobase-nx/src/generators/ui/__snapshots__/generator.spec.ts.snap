// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui generator should run successfully: apps/ui/project.json 1`] = `
"{
  \\"name\\": \\"ui\\",
  \\"$schema\\": \\"../../node_modules/nx/schemas/project-schema.json\\",
  \\"sourceRoot\\": \\"apps/ui/src\\",
  \\"projectType\\": \\"application\\",
  \\"targets\\": {
    \\"build\\": {
      \\"executor\\": \\"@nrwl/webpack:webpack\\",
      \\"outputs\\": [
        \\"{options.outputPath}\\"
      ],
      \\"defaultConfiguration\\": \\"production\\",
      \\"options\\": {
        \\"compiler\\": \\"babel\\",
        \\"outputPath\\": \\"dist/apps/ui\\",
        \\"index\\": \\"apps/ui/src/index.html\\",
        \\"baseHref\\": \\"/\\",
        \\"main\\": \\"apps/ui/src/main.tsx\\",
        \\"polyfills\\": \\"apps/ui/src/polyfills.ts\\",
        \\"tsConfig\\": \\"apps/ui/tsconfig.app.json\\",
        \\"assets\\": [
          \\"apps/ui/src/favicon.ico\\",
          \\"apps/ui/src/assets\\"
        ],
        \\"styles\\": [
          \\"apps/ui/src/styles.css\\"
        ],
        \\"scripts\\": [],
        \\"webpackConfig\\": \\"@nrwl/react/plugins/webpack\\"
      },
      \\"configurations\\": {
        \\"development\\": {
          \\"extractLicenses\\": false,
          \\"optimization\\": false,
          \\"sourceMap\\": true,
          \\"vendorChunk\\": true
        },
        \\"production\\": {
          \\"fileReplacements\\": [
            {
              \\"replace\\": \\"apps/ui/src/environments/environment.ts\\",
              \\"with\\": \\"apps/ui/src/environments/environment.prod.ts\\"
            }
          ],
          \\"optimization\\": true,
          \\"outputHashing\\": \\"all\\",
          \\"sourceMap\\": false,
          \\"namedChunks\\": false,
          \\"extractLicenses\\": true,
          \\"vendorChunk\\": false
        }
      }
    },
    \\"serve\\": {
      \\"executor\\": \\"@nrwl/webpack:dev-server\\",
      \\"defaultConfiguration\\": \\"development\\",
      \\"options\\": {
        \\"buildTarget\\": \\"ui:build\\",
        \\"hmr\\": true
      },
      \\"configurations\\": {
        \\"development\\": {
          \\"buildTarget\\": \\"ui:build:development\\"
        },
        \\"production\\": {
          \\"buildTarget\\": \\"ui:build:production\\",
          \\"hmr\\": false
        }
      }
    },
    \\"lint\\": {
      \\"executor\\": \\"@nrwl/linter:eslint\\",
      \\"outputs\\": [
        \\"{options.outputFile}\\"
      ],
      \\"options\\": {
        \\"lintFilePatterns\\": [
          \\"apps/ui/**/*.{ts,tsx,js,jsx}\\"
        ]
      }
    },
    \\"test\\": {
      \\"executor\\": \\"@nrwl/jest:jest\\",
      \\"outputs\\": [
        \\"coverage/apps/ui\\"
      ],
      \\"options\\": {
        \\"jestConfig\\": \\"apps/ui/jest.config.ts\\",
        \\"passWithNoTests\\": true
      }
    }
  },
  \\"tags\\": []
}
"
`;

exports[`ui generator should run successfully: apps/ui/src/main.tsx 1`] = `
"import { StrictMode } from 'react';
import * as ReactDOM from 'react-dom/client';


import App from './app/app';

const root = ReactDOM.createRoot(document.getElementById('root') as HTMLElement);
root.render(
<StrictMode><App /></StrictMode>
);
"
`;

exports[`ui generator should run successfully: apps/ui/src/styles.css 1`] = `
"@tailwind base;
@tailwind components;
@tailwind utilities;
"
`;

exports[`ui generator should run successfully: package.json 1`] = `
"{
  \\"name\\": \\"test-name\\",
  \\"dependencies\\": {
    \\"@caliobase/caliobase-ui\\": \\"0.1.9\\",
    \\"core-js\\": \\"^3.6.5\\",
    \\"react\\": \\"18.2.0\\",
    \\"react-dom\\": \\"18.2.0\\",
    \\"regenerator-runtime\\": \\"0.13.7\\",
    \\"tslib\\": \\"^2.3.0\\"
  },
  \\"devDependencies\\": {
    \\"@nrwl/cypress\\": \\"14.7.18\\",
    \\"@nrwl/eslint-plugin-nx\\": \\"14.7.18\\",
    \\"@nrwl/jest\\": \\"14.7.18\\",
    \\"@nrwl/linter\\": \\"14.7.18\\",
    \\"@nrwl/react\\": \\"14.7.18\\",
    \\"@nrwl/web\\": \\"14.7.18\\",
    \\"@testing-library/react\\": \\"13.4.0\\",
    \\"@types/jest\\": \\"28.1.1\\",
    \\"@types/node\\": \\"16.11.7\\",
    \\"@types/react\\": \\"18.0.20\\",
    \\"@types/react-dom\\": \\"18.0.6\\",
    \\"@typescript-eslint/eslint-plugin\\": \\"^5.36.1\\",
    \\"@typescript-eslint/parser\\": \\"^5.36.1\\",
    \\"babel-jest\\": \\"28.1.1\\",
    \\"cypress\\": \\"^10.7.0\\",
    \\"eslint\\": \\"~8.15.0\\",
    \\"eslint-config-prettier\\": \\"8.1.0\\",
    \\"eslint-plugin-cypress\\": \\"^2.10.3\\",
    \\"eslint-plugin-import\\": \\"2.26.0\\",
    \\"eslint-plugin-jsx-a11y\\": \\"6.6.1\\",
    \\"eslint-plugin-react\\": \\"7.31.8\\",
    \\"eslint-plugin-react-hooks\\": \\"4.6.0\\",
    \\"jest\\": \\"28.1.1\\",
    \\"jest-environment-jsdom\\": \\"28.1.1\\",
    \\"react-test-renderer\\": \\"18.2.0\\",
    \\"ts-jest\\": \\"28.0.5\\",
    \\"ts-node\\": \\"10.9.1\\"
  }
}
"
`;
